ICU-21169 Add SimpleUnitImpl::getSimpleUnitID() method

hugovdm opened the pull request:
Also swap a UCharsTrie for a BytesTrie, providing us with a set of char* values we can also return through getIdentifier().

<!--
Thank you for your pull request.
Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html#License
-->

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21169
- [X] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

hugovdm commented on icu4c/source/i18n/measunit_extra.cpp:
```diff
@@ -777,6 +775,7 @@ MeasureUnit SingleUnitImpl::build(UErrorCode& status) const {
     return std::move(temp).build(status);
 }
 
+const char *SingleUnitImpl::getIdentifier() const { return gSimpleUnits[index]; }
```
Multi-line: check/change clang-format options...

hugovdm replied:
Reformatted.
The .clang_format option for this is:

    AllowShortFunctionsOnASingleLine: false

younies commented on icu4c/source/i18n/measunit_extra.cpp:
```diff
@@ -230,37 +230,36 @@ UBool U_CALLCONV cleanupUnitExtras() {
+    b.add("per-", INITIAL_COMPOUND_PART_PER, status);
+    b.add("square-", POWER_PART_P2, status);
+    b.add("cubic-", POWER_PART_P3, status);
+    b.add("p2-", POWER_PART_P2, status);
```
I think it is better to have also "p1-"

hugovdm replied:
The specification specifically supports "pow2" to "pow15":
http://unicode.org/reports/tr35/tr35-general.html#Unit_Identifiers

A bigger concern: that specification says "pow#", not "p#"...
I find `pow4-second-square-ampere-per-kilogram-square-meter` in the unitQuantities resource.
We may need to do a s/p/pow/ - but it's also out of scope for this PR.

younies commented on icu4c/source/i18n/measunit_impl.h:
```diff
@@ -32,6 +32,8 @@ struct SingleUnitImpl : public UMemory {
     /** Transform this SingleUnitImpl into a MeasureUnit, simplifying if possible. */
     MeasureUnit build(UErrorCode& status) const;
 
+    const char *getIdentifier() const;
```
Optional: more documentation please

sffc replied:
It would be useful to say that the returned identifier points to static memory and does not need to be cleaned up.

hugovdm replied:
getIdentifier was misleading because this won't return "square-kilometer", even though that would be the identifier for a MeasureUnit containing only this SingleUnit. I've renamed it to "getSimpleUnitID" as inspired by terminology from
https://unicode-org.github.io/cldr/ldml/tr35-general.html#Unit_Identifiers
- and I've improved the documentation.

jefgen commented on icu4c/source/i18n/measunit_extra.cpp:
```diff
@@ -271,14 +273,13 @@ void U_CALLCONV initUnitExtras(UErrorCode& status) {
-    kSerializedUnitExtrasStemTrie = static_cast<char16_t*>(uprv_malloc(numBytes));
-    uprv_memcpy(kSerializedUnitExtrasStemTrie, result.getBuffer(), numBytes);
+    size_t numBytes = result.length();
+    kSerializedUnitExtrasStemTrie = static_cast<char *>(uprv_malloc(numBytes));
```
Technically unrelated to this specific change/PR, but `uprv_malloc`/`malloc` can fail for some platforms (due to out-of-memory conditions) and return `NULL`/`nullptr`.

If you don't mind, I think it would be good to check for `nullptr` here, and set the error code to `U_MEMORY_ALLOCATION_ERROR`. :)

Perhaps something like this:
```c++
if (kSerializedUnitExtrasStemTrie == nullptr) {
    status = U_MEMORY_ALLOCATION_ERROR;
    return;
}
uprv_memcpy(kSerializedUnitExtrasStemTrie, result.data(), numBytes);
```

sffc replied:
Yeah, good catch

hugovdm replied:
Done!

jefgen replied:
Thanks!

hugovdm commented:
Squashing!

hugovdm commented:
Sorry, I somehow managed to not add line breaks to the one-liner function.
880b0e1 added the nullptr check.
701128b added two line-breaks, and hasn't been reviewed, but I'm confident...
I'm boldly squashing again without waiting, to save a round-trip.

hugovdm commented:
My conclusion is that my latest action is indeed highly un-recommended and should be strongly discouraged. What I learned: the Jira-GitHub PR Checker Bot claims "no changes!" - but that's because it was comparing 7b30cd7 to 701128b, with 701128b thereby flying under GitHub PR Radar.

The actions were:
880b0e1 added the nullptr check, and was approved.
I squashed to: bdd2ae8.
I pushed another commit: 701128b.
I squashed to 7b30cd7 not waiting for review.

Probably quite non-trivial to improve automation for this. History can be investigated via links in "[person] force-pushed <...>" log entries. Particularly if someone pushes more than one commit before squashing again... (one has to click on 701128b and then check its parent if one wants to verify there were no other commits between it and bdd2ae8).

sffc commented:
> Probably quite non-trivial to improve automation for this. History can be investigated via links in "[person] force-pushed <...>" log entries. Particularly if someone pushes more than one commit before squashing again... (one has to click on 701128b and then check its parent if one wants to verify there were no other commits between it and bdd2ae8).

This is precisely what an astute reviewer should be doing.
